### ISSUE TITLE
Conditional temporary creation with consumption outside

### ISSUE THREAD

**Issue by jdahm** [GridTools/gt4py#220]
We should warn or error if a temporary is created inside only one branch of a conditional and used outside it. Created from discussion in #193.

**Comment by havogt** [comment]
Check how it relates to #54 (where the check seems to be too strict)

**Comment by havogt** [comment]
The following 2 snippets should produce a descriptive error message, today they randomly fail or run and produce unexpected output depending on backend
```
@gtscript.stencil(backend=backend)
def issue_220(
    in_field: gtscript.Field[dtype], mask: gtscript.Field[bool], out_field: gtscript.Field[dtype]
):
    with computation(FORWARD):
        with interval(...):
            if mask:
                tmp = in_field
            out_field = tmp
```

```
@gtscript.stencil(backend=backend)
def issue_220(
    in_field: gtscript.Field[dtype], mask: bool, out_field: gtscript.Field[dtype]
):
    with computation(FORWARD):
        with interval(...):
            if mask:
                tmp = in_field
            out_field = tmp
```

but the we need to ensure that the following still works
```
@gtscript.stencil(backend=backend)
def issue_220(
    in_field: gtscript.Field[dtype], mask: gtscript.Field[bool], out_field: gtscript.Field[dtype]
):
    with computation(FORWARD):
        with interval(...):
            if mask:
                tmp = in_field
                out_field = tmp
```

**Comment by egparedes** [comment]
In some of the shaping cycles for the new frontend we had some discussions around this topic. I'm posting here a summary of the conclusions because I think it's also relevant for this issue (special thanks to @BenWeber42 for the formal definition and the links):

> Every time a variable is used, it has to be guaranteed (statically) that the variable is initialized at that point. More formally: for each statement, there is a set of initialized variables before and after that statement and only the previously initialized ones can be safely referenced. A simple forward flow analysis can be used to correctly deduce those sets for each statement. If-blocks are treated as a single statement which only initializes variables defined in both branches.

Further reading:
- Definite assignment analysis: https://en.wikipedia.org/wiki/Definite_assignment_analysis
- Data-flow analysis: https://en.wikipedia.org/wiki/Data-flow_analysis

*Note that this issue only deals with `if`-conditions, but a similar issue appears for temporary fields within vertical intervals, since a temporary field could be initialized only in a subset of the vertical range, and thus should not be referenced later from levels outside this range.

**Comment by tehrengruber** [comment]
Fixed by #630. See the description there for more information.
